include/dsp.h

```diff
@@ -447,6 +447,14 @@ struct pmsm_model_f32_s
   float                         iq_int; /* Iq integral part */
 };
 
+/* Average filter */
+
+struct avg_filter_data_s
```

Comment:
I'm not very familiar with options to implement average filter, but my "applied mathematician feeling" plus experience in embedded SW design force me to write next comment.
It is good to implement average filter not only from DSP perspective, but from OOP perspective as well. The `struct avg_filter_data_s` implements average filter "class" and in general its members should be private (that is not obviously possible in C). But each field should have its own meaning. I will try to explain further. I see the average filter to be implemented as a next interface:
```
struct avg_filter_data_s
{
  float avg; /* Current average filtered value */
  float nsample; /* Number of sample participated in averaging */
};

void avg_filter_init(FAR struct avg_filter_data_s *data,
                     float avg, float nsamples);
float avg_filter_proc(FAR struct avg_filter_data_s *data, float x);
float avg_filter_get_average(FAR struct avg_filter_data_s *data);
float avg_filter_get_nsamples(FAR struct avg_filter_data_s *data);

...

float avg_filter_proc(FAR struct avg_filter_data_s *data, float x)
{
  float alpha;
  float new_avg;

  LIBDSP_DEBUGASSERT(data != NULL);

  alpha = data->nsample / (data->nsample + 1.0f);

  new_avg = (alpha * data->avg) + ((1.0f - alpha) * x);
  data->nsample += 1.0f;

  data->avg = new_avg;

  return data->avg;
}
```
The `avg_filter_init()` is obviously a constructor while naming private member not "prev_avg", but "avg" gives an option to implement `avg_filter_get_average()` to return the currently accumulated average value and that formally will not be possible with `prev_avg` because otherwise interface should be named `avg_filter_get_prev_average`. I will explain with an example:
```
void my_dsp_func(float data[], unsigned data_count)
{
  struct avg_filter_data_s filter;
  unsigned j;

  avg_filter_init(&filter, 0.0f, 0.0f);

  for (j = 0; j < data_count, j++)
    {
      avg_filter_proc(&filter, data[j]);
    }
...
  apply_averaged(avg_filter_get_average(&filter));
...
}
```
With such approach the average filter "class" is holding all the need state information and user is not forced to use auto variable to store result from each `avg_filter()` call.
Also `k` in current version represents the number of samples used during average calculation so initially it is 0 (no samples) and is incremented by 1 with each call of `avg_filter()` while having `k` ignited to `1` has no other meaning that reference to a particular part in video that explained the algorithm operation.

Reply:
@pkarashchenko I don't know I think the other way with only a init() and avg_filter() was simpler. Now we are forcing user to call three functions and we lost the mathematical meaning (keep prev_avg variables). Maybe it's impossible to please two gods! :-)

Reply:
what we like to control in any filtering process that's the rolling window size, especially in lowpass filtering ; i.e that's a noise reducer.

Reply:
Hi @mu578 yes it is on radar too (moving average filter).

Reply:
BTW, since we are having "too many" ideas here, please @raiden00pl, @pkarashchenko, @xiaoxiang781216, @mu578, @hartmannathan could you please approve or reject this PR to let it be merged or closed.

Reply:
@accassis myself I am not against it, but I don't see much of interest. If one implements primitives in any type/field API; it must have a pertinent and practical usage.

If any DSP primitives are implemented, it must relay on existing (e.g what others do) ; personally, I like `scipy.signal` as a reference and compare base.

Why? because as a former full-time DSP software engineer it speaks to me by design.

I am no fan of `stat like per sample processed next`, I am a practical guy in mind and in usage, we work with processing windows. Now, I am always open to discussion.

Reply:
Thank you for your suggestion @mu578! I also like to follow some standard, but AFAIK there is not good C standard dsp signal APIs and our current libdsp is not following any existing standard, correct @raiden00pl ?

Anyway the NuttX libdsp is very simple and intuitive to use.  scopy.signal fall shorter here, see an example about moving average: https://stackoverflow.com/questions/13728392/moving-average-or-running-mean/43200476#43200476

Should we follow CMSIS-DSP? https://schaumont.dyn.wpi.edu/ece4703b21/lecture7.html

I don't think it is a good idea, because vendors that are ARM's competitor will not like it.

So, I suggest KISS methodology like libdsp seems to be following. If in the future POSIX defines some standard to DSP library we could adopt it.

Reply:
"there is not good C standard dsp signal APIs" , there is armadillo but it is bloated with c++ ; scipy.signal is very C like designed ; so it is not tremendously difficult to follow ; I have two references which I usually clone in plain C; matlab and scipy.signal ; then I can compare results and if algorithm is right.